client/question/services/FeedbackGeneratorService.js

```diff
@@ -263,33 +328,70 @@ tie.factory('FeedbackGeneratorService', [
      * @param {string|null} outputFunctionName If needed, is string of the
      *    name for the function needed to process/format the user's output.
      * @param {CorrectnessTest} TestCase
+     * @param {number} testCaseIndex
      * @param {*} observedOutput Actual output for running user's code.
+     * @param {string} testSuiteId
      * @returns {Feedback}
      * @private
      */
     var _getCorrectnessTestFeedback = function(
-      outputFunctionName, testCase, observedOutput) {
+        outputFunctionName, testCase, testCaseIndex, observedOutput,
+        testSuiteId) {
+      var testCaseStateKey = testSuiteId + '-' + testCaseIndex;
       var allowedOutputExample = testCase.getAnyAllowedOutput();
       var feedback = FeedbackObjectFactory.create(
         FEEDBACK_CATEGORIES.INCORRECT_OUTPUT_FAILURE);
-      // TODO(eyurko): Add varied statements for when code is incorrect.
-      feedback.appendTextParagraph('Your code produced the following result:');
-      if (outputFunctionName) {
-        feedback.appendCodeParagraph(
-          'Input: ' + _jsToHumanReadable(testCase.getInput()) + '\n' +
-          ('Output (after running ' + outputFunctionName + '): ' +
-            _jsToHumanReadable(observedOutput)));
-        feedback.appendTextParagraph(
-          'However, the expected output of ' + outputFunctionName + ' is:');
+      if (testSuiteId == lastTestSuiteId) {
+        if (correctnessTestStates[testCaseStateKey] ==
+            CORRECTNESS_TEST_INPUT_DISPLAYED &&
+            testSuiteId != SAMPLE_INPUT) {
```

Comment:
Having SAMPLE_INPUT in here is a bit convoluted. I recommend the following:

- At the outset, make sure the state is correct:
  - Check whether the state doesn't exist, if so, initialize it.
  - Check whether the suite ID is sample input; if it is then set the current state to be OBSERVED_OUTPUT_DISPLAYED.

Now, you can just act based only on the state, and do `if (state == SUCCESS) {...} else if (state == OUTPUT_ENABLED) {...} else if (state == EXPECTED_OUTPUT_DISPLAYED) {...} else if (state == INPUT_DISPLAYED) {...} else {throw Error('...')}`. This helps split up the logic and makes it easier to maintain this in the future.

Reply:
Let me think about this. I'm not sure setting OBSERVED_OUTPUT_DISPLAYED on SAMPLE_INPUT would work, given it would bypass the hinting phase. I'll think about how to make this simpler and more readable.

Reply:
Updated in latest commit.